src/main/resources/project/server/deployApplication.pl

```diff
@@ -0,0 +1,446 @@
+            if @specified_disabled_server_groups;
+
+        my %all_specified_server_groups_hash = map {$_ => 1} (@specified_enabled_server_groups,
+            @specified_disabled_server_groups);
+
+        my @non_existing_server_groups = grep {
+            !$json_server_groups_discovery->{result}->{$_}
+        } (@specified_enabled_server_groups, @specified_disabled_server_groups);
+        $jboss->bail_out("Specified non existing server group(s): @non_existing_server_groups. Please add server groups before deploying to them.")
+            if @non_existing_server_groups;
+
+        my @missing_server_groups = grep {!$all_specified_server_groups_hash{$_}} @all_server_groups;
+
+        my @missing_enabled_server_groups = grep {
+            $json_server_groups_discovery->{result}->{$_}->{deployment}
+                && $json_server_groups_discovery->{result}->{$_}->{deployment}->{$expected_deployment_name}
+                && $json_server_groups_discovery->{result}->{$_}->{deployment}->{$expected_deployment_name}->{enabled}
+        } @missing_server_groups;
+        $jboss->log_info("Deployment '$expected_deployment_name' is assigned and enabled on the server group(s) which are not in enabled/disabled list: @missing_enabled_server_groups. Leave the deployment enabled on this server group(s).")
+            if @missing_enabled_server_groups;
+
+        # needed for WFCORE-2939 workaround
+        my @missing_disabled_server_groups = grep {
+            $json_server_groups_discovery->{result}->{$_}->{deployment}
+                && $json_server_groups_discovery->{result}->{$_}->{deployment}->{$expected_deployment_name}
+                && !$json_server_groups_discovery->{result}->{$_}->{deployment}->{$expected_deployment_name}->{enabled}
+        } @missing_server_groups;
+        $jboss->log_info("Deployment '$expected_deployment_name' is assigned and disabled on the server group(s) which are not in enabled/disabled list: @missing_disabled_server_groups. WFCORE-2939 workaround to be applied to make sure that we leave the deployment disabled on this server group(s).")
+            if @missing_disabled_server_groups;
+
+        my @specified_server_groups_where_deployment_was_enabled = grep {
+            $all_specified_server_groups_hash{$_}
+                && $json_server_groups_discovery->{result}->{$_}->{deployment}
+                && $json_server_groups_discovery->{result}->{$_}->{deployment}->{$expected_deployment_name}
+                && $json_server_groups_discovery->{result}->{$_}->{deployment}->{$expected_deployment_name}->{enabled}
+        } @all_server_groups;
+        $jboss->log_info("FYI.. Deployment '$expected_deployment_name' is assigned and enabled on the server group(s) which are in enabled/disabled list: @specified_server_groups_where_deployment_was_enabled")
+            if @specified_server_groups_where_deployment_was_enabled;
+
+        my @specified_server_groups_where_deployment_was_disabled = grep {
```

Comment:
Why do we need a bunch of these greps, if the logic is similar?

Reply:
there is different conditions in grep functions for each array. Some of them has almost the same, but "!" sign makes difference.